### ISSUE TITLE
Some times a module is accepted "too early" in the dependency tree

### ISSUE THREAD

**Issue by mnzaki** [noppa/ng-hot-reload#4]
I'm using the webpack `ng-hot-reload-loader` and using the recommended method of applying it to all '.js' files.

Depending on project layout, this does not necessarily always work.
I have some sub directories that contain collections of components and an index file that looks something like this:
```
let compRegex = /\.*\/(.*)\.component\.js$/,
  // NOTE: compRegex is copied as the last argument because when using a
  // variable webpack can't parse the context
  reqContext = require.context('./', true, /.*\.component\.js$/),
  ngMod = angular.module('some.components', []);

reqContext.keys().forEach(compPath => {
  let comp = reqContext(compPath).default,
    name = compRegex.exec(compPath)[1];
  ngMod.component(name, comp);
});

export default ngMod.name;
```
What this does is essentially look for all `.component.js` files and register them all on an angular module. The `.component.js` files simply export a component object (and does not call `angular.component`)

The problem is that `ng-hot-reload-loader` will wrap the `.component.js` files themselves and call `module.hot.accept` inside them, so the index file is never re-evaluated (so the component registration is never called during a hot reload).

A suggested solution would be to keep track of whether or not `angular.module` was ever called, and if it wasn't then we don't accept the hot reload. This will cause the reload to bubble up the module dependency chain, and so will reach the index file. In the index file `angular.module` will be called and the module can be accepted.

I see that such a similar check mechanism is used (`updateId`) in some places, but another check is needed in the `ng-hot-reload-loader` wrapper, specifically in the `finally { ... }` clause that has `module.hot.accept`, it would have to check here if `angular.module` was ever called, and only accept the hot reload if it is called.

**Comment by mnzaki** [comment]
I have also tried wrapping only certain parts with the loader, but that runs into a different issue:
The decorated `angular` is only local to the specific wrapped file, when a dependency is required then 'angular' inside that dependency's code refers to the global angular variable (which is not decorated) and thus doesn't work as expected.

Is there a specific reason why the global angular object is not directly changed? What about changing the global object then returning it to normal after the source is evaluated?

**Comment by noppa** [comment]
Thanks for the issue and the suggestions! 
I haven't worked on this project for a while, so I'll need to refresh my memory a bit before I can give good answers to your suggestions.

>Is there a specific reason why the global angular object is not directly changed? What about changing the global object then returning it to normal after the source is evaluated?

I didn't want to modify the global object because some libraries and other code that the user doesn't want to get hot-reloaded might still want to use the original methods directly. I'm actually not using Webpack myself at the moment, supporting it was kind of an "extra feature", so I'll need to play around with this issue a bit before I can say much more about that. If you happen to have a repo where this issue could be reproduced, it would definitely save some time.

Anyways, thanks for taking interest in the project, I'll try to find a bit more time for it now that there's at least one other person using it 😅  
I'll come back to this tomorrow or in a few days.

**Comment by mnzaki** [comment]
I played around with different ideas to solve the issue. Thing is, supporting every possible use case (especially differences in project layouts) is probably going to be very problematic....

Yesterday I tested changing the global object and restoring it at the end of the loader code. I did this using the loader only on the entry point of the bundle I would like to support hot reloading for (instead of globally on all `*.js`). This actually kinda worked except it breaks when combined with `ng-template-loader` (which I think is fairly standard for any angularjs + webpack installation). The issue is that `ng-template-loader` injects and uses the `$templateCache` in a `run` block on the builtin `ng` module. And of course because the global object is hijacked, the `ng` module is requested and tracked for changes and `ng-hot-reload` decides that the reload wasn't handled appropriately and refreshes the page :disappointed:

What do you think about `run` blocks and how they should be handled in a hot reload? In this specific case it would actually be totally fine to just run the block, but I'm not sure if that generalizes.

I had been looking for HMR for AngularJS a long time ago. I finally got fed up enough yesterday and decided to implement it on my own but thought "hey just one more google search, just in case someone already did it" and lo and behold :D thank you for making this!

I will try to set up a little project to demonstrate some of these issues/ideas, hopefully make `ng-hot-reload` a little more versatile!

**Comment by mnzaki** [comment]
I just found a very interesting commit that landed in angular 1.6.7: angular/angular.js@34237f929927295392bbb1a600e78cbde581839a

and the related issue angular/angular.js#11015 (which started in Feb 2015 :sweat_smile:)

**Comment by noppa** [comment]
@mnzaki This should be fixed now (in version 2.0.0).  
Now the loader won't call `module.hot.accept` if there are no calls to `.component` `.directive` or `.controller` in that file.  
Let me know if the issue still persists after upgrading, thanks!
